vars/xsDeploy.groovy

```diff
@@ -0,0 +1,329 @@
+import com.sap.piper.JenkinsUtils
+
+import static com.sap.piper.Prerequisites.checkScript
+
+import com.sap.piper.BashUtils
+import com.sap.piper.ConfigurationHelper
+import com.sap.piper.GenerateDocumentation
+import com.sap.piper.Utils
+
+import groovy.transform.Field
+
+import hudson.AbortException
+
+@Field String STEP_NAME = getClass().getName()
+
+@Field Set GENERAL_CONFIG_KEYS = STEP_CONFIG_KEYS
+
+@Field Set STEP_CONFIG_KEYS = [
+    'action',
+    'apiUrl',
+    'credentialsId',
+    'deploymentId',
+    'deployIdLogPattern',
+    'deployOpts',
+    /** A map containing properties forwarded to dockerExecute. For more details see [here][dockerExecute] */
+    'docker',
+    'loginOpts',
+    'mode',
+    'mtaPath',
+    'org',
+    'space',
+    'xsSessionFile',
+]
+
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+enum DeployMode {
+    DEPLOY,
+    BG_DEPLOY,
+    NONE
+
+    String toString() {
+       name().toLowerCase(Locale.ENGLISH).replaceAll('_', '-')
+    }
+}
+
+enum Action {
+    RESUME,
+    ABORT,
+    RETRY,
+    NONE
+
+    String toString() {
+       name().toLowerCase(Locale.ENGLISH)
+    }
+}
+
+/**
+  * Performs an XS deployment
+  */
+@GenerateDocumentation
+void call(Map parameters = [:]) {
+
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: parameters) {
+
+        def utils = parameters.juStabUtils ?: new Utils()
+
+        final script = checkScript(this, parameters) ?: this
+
+        ConfigurationHelper configHelper = ConfigurationHelper.newInstance(this)
+            .loadStepDefaults()
+            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+            .mixinStageConfig(script.commonPipelineEnvironment, parameters.stageName?:env.STAGE_NAME, STEP_CONFIG_KEYS)
+            .addIfEmpty('mtaPath', script.commonPipelineEnvironment.getMtarFilePath())
+            .addIfEmpty('deploymentId', script.commonPipelineEnvironment.xsDeploymentId)
+            .mixin(parameters, PARAMETER_KEYS)
+
+        Map config = configHelper.use()
+
+        DeployMode mode = config.mode
+
+        if(mode == DeployMode.NONE) {
+            echo "Deployment skipped intentionally. Deploy mode '${mode.toString()}'."
+            return
+        }
+
+        Action action = config.action
+
+        if(mode == DeployMode.DEPLOY && action != Action.NONE) {
+            error "Cannot perform action '${action.toString()}' in mode '${mode.toString()}'. Only action '${Action.NONE.toString()}' is allowed."
+        }
+
+        boolean performLogin  = ((mode == DeployMode.DEPLOY) || (mode == DeployMode.BG_DEPLOY && !(action in [Action.RESUME, Action.ABORT])))
+        boolean performLogout = ((mode == DeployMode.DEPLOY) || (mode == DeployMode.BG_DEPLOY && action != Action.NONE))
+
+        boolean sessionExists = fileExists file: config.xsSessionFile
+
+        if( (! performLogin) && (! sessionExists) ) {
```

Comment:
The session is represented by a file `.xsconfig` in the home directory of the user. Since we have several calls to `xs`, each fron another docker container we have to remember this file in the workspace. When we perform the login the file gets copied into the workspace. For subsequent actions we copy the file from the workspace into the user home directory before calling `xs`.

❗️  The file contains a bearer token. Between the login and the logout (which happens in case of a bg-deploy later after resume/abort) it is possible to hijack the session by copying this file. This is actually bad. In case someone has a better idea how to store the session, please raise your voice ...

Reply:
As I learned in the meantime there is the flag `--context-file` and also the environment variable 'XSCLIENT_CONTEXTFILE'. With using one of both approaches we can drop the copy actions between the workspace and the user home directory. We have still the issue with having the context file in the workspace (bearer token), but at least there is no need anymore for the copy actions.